### PR TITLE
Create API.md

### PR DESCRIPTION
**Pull request by AdrienCorenflos**
Added draft of API proposal

### CODE REVIEW

File: design-notes/API.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
+# API design notes
+
+The goal of these notes is to describe (and agree on!) the common API/style/code that would be the foundation of the methods. 
+
+Several principles guide the design:
+- Flexibility/reusability: a function should require exactly what it needs to run. 
+- Pure functional: no side effect
+- Typed without overloading: for readability, not for logic.
+
```
Comment (murphyk) [inline design-notes/API.md:9]: I agree with these principles.

File: design-notes/API.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+## Gaussian approximated methods
+
+### Model definition
+
+The main user defined function will be the way to compute (or approximate) conditional mean and covariances
+
```
Comment (murphyk) [inline design-notes/API.md:16]: A key question is: do we define our own MVN 'model', or do we leverage distrax?
They recently added support for full covariance MVNs, using cholesky decomposition:
https://github.com/deepmind/distrax/blob/master/distrax/_src/distributions/mvn_full_covariance.py
They also support low rank versions. I am inclined to use their primitives.

File: design-notes/API.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+```python
+def make_conditional_model(f: Callable[[np.ndarray, np.ndarray, ...], np.ndarray], integration_method, sqrt:bool) -> Union[CovarianceModel, CholeskyModel]:
+    # For example here we suppose that `q` is a MVNCov
+    if sqrt:
+        @covariance_model
+        def mean_cov(m_t, cov_t, *args, **kwargs):
+            def mean(x_t):
+                # Something like this I guess, but we can iterate.
+                F, bias, _ = linearize(lambda q_: f(x_t, q_, *args, **kwargs), integration_method, MVNCov(m_t, cov_t))
+                return F @ m_t + bias
```
Comment (murphyk) [inline design-notes/API.md:91]: A trivial thing we should agree on is notation.
Sarkka (and others) use
```
x(k) = F x(k-1) + G u(k-1) + noise(0, Q)
y(k)  = H x(k) + noise(0,R)
```
However, in much of ML (eg VAE literature), x is an observed variable, not latent.
Also, H is often the Hessian. I would prefer
```
z(t) = A z(t-1) + B u(t-1) + noise(0, Q)
y(t)  = C z(t) + D u(t) + noise(0,R)
```
Or we could be even more explicit and write
```
state(t) = trans_mat * state(t-1) + input_state_mat * input(t-1) + noise(0, state_noise_cov)
obs(t)  = obs_mat * state(t) + input_obs_mat * input(t) + noise(0, obs_noise_cov)
```

File: design-notes/API.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+```
+
+Finally, the user can also pass in LGSSMs directly to a public Kalman API beside the non-linear extensions. This will be done in a completely explicit way, where the user will feed in all matrices and say if they want the sqrt or standard form. That is,
+
+```python
+def kalman_filter(ys: np.ndarray, F: np.ndarray, Q: Union[MVNCov, MVNChol], H, R:Union[MVNCov, MVNChol], x0: Union[MVNCov, MVNChol], propagate_first:bool = False):
+    ...
```
Comment (murphyk) [inline design-notes/API.md:98]: Things like `Union[MVNCov, MVNChol]` become hard to read. Maybe we need aliases like
```
MVNDist = Union[MVNCov, MVNChol]
```

File: design-notes/API.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            def mean(x_t):
+                # Something like this I guess, but we can iterate.
+                F, bias, _ = linearize(lambda q_: f(x_t, q_, *args, **kwargs), integration_method, MVNCov(m_t, cov_t))
+                return F @ m_t + bias
+```
+
+Finally, the user can also pass in LGSSMs directly to a public Kalman API beside the non-linear extensions. This will be done in a completely explicit way, where the user will feed in all matrices and say if they want the sqrt or standard form. That is,
+
+```python
+def kalman_filter(ys: np.ndarray, F: np.ndarray, Q: Union[MVNCov, MVNChol], H, R:Union[MVNCov, MVNChol], x0: Union[MVNCov, MVNChol], propagate_first:bool = False):
```
Comment (murphyk) [inline design-notes/API.md:97]: We should allow all the parameters to be optionally parameterized by time, to handle non-stationary models.
This is what we did in https://github.com/probml/JSL/tree/main/jsl/lds:

```
@dataclass
class LDS:
    A: chex.Array
    C: Union[chex.Array, Callable]
    Q: chex.Array
    R: chex.Array
    mu0: chex.Array
    Sigma0: chex.Array

    def observation_matrix(self, t: int):
        if callable(self.C):
            return self.C(t)
        else:
            return self.C


def kf(params, obs, ...):
  for t in range(T):
       Ct = params.observations(t)


```
I know you don't like OOP, but the above is still functionally pure, since dataclass is immutable.

File: design-notes/API.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+Some additional considerations can be handled in the future, such as 
+1. Masked arrays for (partially) missing observations.
+2. Parametrized models: e.g., `F = F(t)` that can be optimized in memory.
+3. Diagonal noise models that can be solved more efficiently.
+
+
+On the other hand, non-linear models will follow a similar API with the additional constraint that the linearization states can be added:
+
+```python
+def non_linear_kalman_filter(ys: np.ndarray, F: Union[np.ndarray, CovarianceModel, CholeskyModel], Q: Union[MVNCov, MVNChol], H: Union[np.ndarray, CovarianceModel, CholeskyModel], R:Union[MVNCov, MVNChol], x0: Union[MVNCov, MVNChol], linearization_method: Union[Callable, Tuple[Callable, Callable]], linearization_states: Optional[Union[MVNChol, MVNCov]] = None, propagate_first: bool = False, parallel: bool = False, return_log_likelihood: bool=True) -> Tuple[Union[MVNCov, MVNChol], Optional[float]]:
```
Comment (murphyk) [inline design-notes/API.md:120]: I don't understand the type signature for F and H.

**Comment by murphyk** [comment]
Instead of vanilla python dataclass, we might want to use https://github.com/brentyi/jax_dataclasses, which plays nicely with jax pytrees.

**Comment by murphyk** [comment]
I mis-spoke. Dataclasses are mutable (by default). We might want to use NamedTuple instead, eg 
```
Class LGSSM(NamedTuple):
     trans_mat: jnp.array
    ...
```
A similar idiom is used in [JaxOpt](https://github.com/google/jaxopt/blob/main/jaxopt/_src/base.py#L48)
and other libraries.

**Comment by murphyk** [comment]
BTW, we will also need a representation for Gaussians in information  form, as well as methods to convert to/from moment and canonical form. Update rules are given in these papers

H. Loeliger, J. Dauwels, J. Hu, S. Korl, L. Ping, and F. R. Kschischang, “The Factor Graph Approach to Model-Based Signal Processing,” Proc. IEEE, vol. 95, no. 6, pp. 1295–1322, Jun. 2007, doi: 10.1109/JPROC.2007.896497. [Online]. Available: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.408.6704&rep=rep1&type=pdf

C. Herzog né Hoffmann, E. Petersen, and P. Rostalski, “Iterative Approximate Nonlinear Inference via Gaussian Message Passing on Factor Graphs,” IEEE Control Systems Letters, vol. 3, no. 4, pp. 978–983, Oct. 2019, doi: 10.1109/LCSYS.2019.2919260. [Online]. Available: http://dx.doi.org/10.1109/LCSYS.2019.2919260
